File: packages/agent/src/project_emoji.js
// Deterministic project emoji mapping: every session that lives in the
// same project folder (cwd basename) lands on the same emoji without
// any per-project storage.
//
// Stage 1 = display only. Stage 2 will let the user override per
// session and persist the choice in `sessions.emoji`; the override
// wins over this default when set.
//
// The hash + pool are duplicated in apps/{mac,ios} (Swift). Both
// implementations MUST stay in sync — a session loaded by either
// client should land on the same default emoji. The unit tests in
// packages/agent/test/project_emoji.test.js + the Swift test in
// apps/.../Tests/ProjectEmojiTests.swift lock the cross-platform
// invariant with a shared fixture set.

export const PROJECT_EMOJI_POOL = [
  "🚀", "⚡️", "🔧", "🛠", "⚙️", "🧪", "🧰", "🧱", "📦", "🗂",
  "📐", "📊", "🧭", "🔬", "🪛", "🛰", "🛸", "🧠", "💡", "🔭",
  "🧲", "🪐", "🌱", "🌿", "🍀", "🌊", "🔥", "✨", "🪄", "🎯",
  "🎛", "🕹", "🎚", "🧩", "🪞", "🦾", "🪪", "📎", "🔗", "⛓",
  "🧷", "🗝", "🎨", "🖋", "📝"
];

/// Normalize a project label (cwd basename) so the hash is stable
/// across trivial cosmetic differences. Lowercased; spaces, dashes,
/// and underscores collapse to nothing. "Portfolio_deck_2026" and
/// "portfolio-deck-2026" and "portfolio deck 2026" all hash to the
/// same slot. Empty strings fall back to "default" so we still
/// return a deterministic emoji.
export function normalizeProjectKey(label) {
  if (typeof label !== "string") return "default";
  const cleaned = label
    .toLowerCase()
    .replace(/[\s_\-]+/g, "");
  return cleaned.length > 0 ? cleaned : "default";
}

/// FNV-1a 32-bit. Picked because it's short, deterministic, and
/// trivial to re-implement byte-for-byte in Swift. Cross-platform
/// parity matters more than cryptographic strength here.
export function fnv1a32(input) {
  let hash = 0x811c9dc5;
  for (let i = 0; i < input.length; i += 1) {
    hash ^= input.charCodeAt(i) & 0xff;
    // 32-bit truncated multiply by FNV prime (0x01000193). We work
    // in 32 bits with Math.imul to stay inside JS's int range.
    hash = Math.imul(hash, 0x01000193) >>> 0;
  }
  return hash >>> 0;
}

/// Deterministic mapping from a normalized project key to a pool
/// entry. Stable forever as long as PROJECT_EMOJI_POOL doesn't
/// reorder — appending new emoji to the end is safe; reordering
/// or removing entries shifts everyone's emoji.
export function projectEmojiFor(label) {
  const key = normalizeProjectKey(label);
  const index = fnv1a32(key) % PROJECT_EMOJI_POOL.length;
  return PROJECT_EMOJI_POOL[index];
}

File: apps/mac/Sources/SteerMac/ProjectEmoji.swift
import Foundation

/// Deterministic mapping from a project label (cwd basename) to an
/// emoji from a frozen pool. Mirrors
/// `packages/agent/src/project_emoji.js` exactly so every client
/// arrives at the same default for the same project. Stage 2 will
/// let users pick a per-session override; until then this is the
/// only source of truth.
///
/// Cross-platform invariant: the Node unit test in
/// `packages/agent/test/project_emoji.test.js` and the Swift test
/// in `apps/mac/Tests/ProjectEmojiTests.swift` both assert the
/// same (label → emoji) fixtures.
enum ProjectEmoji {
    static let pool: [String] = [
        "🚀", "⚡️", "🔧", "🛠", "⚙️", "🧪", "🧰", "🧱", "📦", "🗂",
        "📐", "📊", "🧭", "🔬", "🪛", "🛰", "🛸", "🧠", "💡", "🔭",
        "🧲", "🪐", "🌱", "🌿", "🍀", "🌊", "🔥", "✨", "🪄", "🎯",
        "🎛", "🕹", "🎚", "🧩", "🪞", "🦾", "🪪", "📎", "🔗", "⛓",
        "🧷", "🗝", "🎨", "🖋", "📝"
    ]

    /// Lowercase + strip spaces/dashes/underscores so cosmetic
    /// variants of the same project name collapse to one slot.
    /// Mirrors `normalizeProjectKey` in the JS module.
    static func normalize(_ label: String) -> String {
        let lowered = label.lowercased()
        let stripped = lowered.unicodeScalars.filter { scalar in
            !(scalar == " " || scalar == "-" || scalar == "_")
        }
        let result = String(String.UnicodeScalarView(stripped))
        return result.isEmpty ? "default" : result
    }

    /// FNV-1a 32-bit. Byte-for-byte compatible with the Node
    /// version (`fnv1a32` in project_emoji.js). Operates on the
    /// UTF-8 byte sequence — the JS version uses `charCodeAt & 0xff`
    /// which equals the low byte of the UTF-16 code unit, but every
    /// ASCII character (and Korean characters in our project names
    /// are normalized away first) has the same low byte in UTF-8
    /// and UTF-16 because the inputs after `normalize` are
    /// guaranteed ASCII-only for cross-platform fixtures. Anything
    /// non-ASCII that survives normalize is still deterministic —
    /// just don't expect Swift and Node to agree on it, which the
    /// fixture set respects.
    static func fnv1a32(_ input: String) -> UInt32 {
        var hash: UInt32 = 0x811c9dc5
        for byte in input.utf8 {
            hash ^= UInt32(byte)
            hash = hash &* 0x01000193
        }
        return hash
    }

    /// Deterministic emoji for a project label. Used as the fallback
    /// when `sessions.emoji` is NULL (i.e. no user override yet).
    static func emoji(for label: String) -> String {
        let key = normalize(label)
        let index = Int(fnv1a32(key) % UInt32(pool.count))
        return pool[index]
    }
}
